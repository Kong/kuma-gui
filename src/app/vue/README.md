---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

