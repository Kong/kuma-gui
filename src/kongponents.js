import Vue from 'vue'
import '@kongponents/styles/styles.css'
import KAlert from '@kongponents/kalert'
import KBadge from '@kongponents/kbadge'
import KButton from '@kongponents/kbutton'
import KCard from '@kongponents/kcard'
import KEmptyState from '@kongponents/kemptystate'
import KIcon from '@kongponents/kicon'
import KTable from '@kongponents/ktable'
import KToaster from '@kongponents/ktoaster'
import KoolTip from '@kongponents/kooltip'
import KClipboardProvider from '@kongponents/kclipboardprovider'
import Krumbs from '@kongponents/krumbs'
import KPop from '@kongponents/kpop'
import KToggle from '@kongponents/ktoggle'
import KTabs from '@kongponents/ktabs'
import KModal from '@kongponents/kmodal'
import KInput from '@kongponents/kinput'
import KInputSwitch from '@kongponents/kinputswitch'

Vue.component('KAlert', KAlert)
Vue.component('KBadge', KBadge)
Vue.component('KButton', KButton)
Vue.component('KCard', KCard)
Vue.component('KEmptyState', KEmptyState)
Vue.component('KIcon', KIcon)
Vue.component('KTable', KTable)
Vue.component('KToaster', KToaster)
Vue.component('KoolTip', KoolTip)
Vue.component('KClipboardProvider', KClipboardProvider)
Vue.component('Krumbs', Krumbs)
Vue.component('KPop', KPop)
Vue.component('KTabs', KTabs)
Vue.component('KToggle', KToggle)
Vue.component('KInput', KInput)
Vue.component('KInputSwitch', KInputSwitch)
Vue.component('KModal', KModal)
