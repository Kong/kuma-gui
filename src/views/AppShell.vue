<template>
  <div>
    <GlobalHeader />
    <div class="main-content-container">
      <AppSidebar />

      <main class="main-content">
        <NotificationManager />
        <OnboardingNotification v-if="showOnboarding" />
        <BreadcrumbsMenu />
        <transition
          mode="out-in"
          name="fade"
        >
          <router-view :key="routeKey" />
        </transition>
      </main>
    </div>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'

import GlobalHeader from '@/components/Global/GlobalHeader.vue'
import AppSidebar from '@/components/Sidebar/AppSidebar.vue'
import NotificationManager from '@/components/NotificationManager/NotificationManager.vue'
import BreadcrumbsMenu from '@/components/BreadcrumbsMenu.vue'
import OnboardingNotification from '@/components/NotificationManager/components/OnboardingNotification.vue'

export default {
  name: 'AppShell',
  components: {
    BreadcrumbsMenu,
    AppSidebar,
    NotificationManager,
    OnboardingNotification,
    GlobalHeader,
  },
  computed: {
    ...mapGetters({
      showOnboarding: 'onboarding/showOnboarding',
    }),

    /**
     * The `router-view`’s `key` attribute value.
     *
     * Is always set to `'default'` (i.e. will never trigger an explicit re-render via Vue’s `key` mechanism).
     * However, in some scenarios, we want Vue to re-render a route’s components
     * (e.g. `src/views/Policies/PolicyView.vue` which is used by some dozen policy routes).
     */
    routeKey() {
      if (this.$route.meta.shouldReRender) {
        return this.$route.path
      }

      return 'default'
    },
  },
}
</script>
