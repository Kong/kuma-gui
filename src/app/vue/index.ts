import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
