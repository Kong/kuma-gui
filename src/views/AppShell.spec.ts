import { createStore } from 'vuex'
import { createRouter, createWebHashHistory } from 'vue-router'
import { flushPromises, RouterLinkStub } from '@vue/test-utils'
import { render } from '@testing-library/vue'
import { KAlert, KBadge, KButton, KIcon, KModal, KPop } from '@kong/kongponents'

import AppShell from './AppShell.vue'
import TestComponent from '@/testUtils/TestComponent.vue'

const router = createRouter({
  history: createWebHashHistory(),
  routes: [
    {
      path: '/',
      name: 'home',
      component: TestComponent,
    },
    {
      path: '/overview',
      name: 'global-overview',
      component: TestComponent,
    },
    {
      path: '/diagnostics',
      name: 'diagnostics',
      component: TestComponent,
    },
  ],
})

const store = createStore({
  modules: {
    config: {
      namespaced: true,
      getters: {
        getStatus: () => undefined,
        getTagline: () => 'Kuma',
        getVersion: () => undefined,
      },
    },
    onboarding: {
      namespaced: true,
      getters: {
        showOnboarding: () => false,
      },
    },
    sidebar: {
      namespaced: true,
      state: {
        insights: {},
      },
    },
    notifications: {
      namespaced: true,
      getters: {
        amountOfActions: () => undefined,
      },
    },
  },
  state: {
    policies: [],
    meshes: [{}],
  },
})

describe('AppShell.vue', () => {
  it('renders component with route', async () => {
    const { container } = render(AppShell, {
      global: {
        plugins: [router, store],
        components: {
          KAlert,
          KBadge,
          KButton,
          KIcon,
          KModal,
          KPop,
        },
        stubs: {
          'router-link': RouterLinkStub,
        },
      },
    })

    await flushPromises()

    expect(container).toMatchSnapshot()
  })
})
