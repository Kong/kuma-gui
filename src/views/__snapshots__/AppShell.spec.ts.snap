// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppShell.vue renders component with route 1`] = `
<div>
  <div>
    <header
      class="main-header px-4 py-1 flex justify-between items-center"
    >
      <div
        class="py-1 md:py-0 flex justify-between items-center"
      >
        <a
          class="logo"
        >
          <img
            alt="Kuma Logo"
          />
        </a>
        <div
          class="my-0 mx-6 upgrade-check-wrapper"
        >
          <div
            class="upgrade-check"
          >
            <div
              class="warning small alert none k-alert"
              data-v-53188520=""
              role="alert"
            >
              <!---->
              <div
                class="warning k-alert-action ml-3"
                data-v-53188520=""
              >
                <!---->
                <!---->
              </div>
              <!---->
              <!---->
              <div
                class="k-alert-msg-text"
                data-v-53188520=""
              >
                <div
                  class="k-alert-msg"
                  data-v-53188520=""
                >
                  
                  <div
                    class="alert-content"
                  >
                    <div>
                      Kuma update available 
                    </div>
                    <div>
                      <a
                        class="small rounded primary k-button warning-button"
                        data-v-29036f3a=""
                        href="https://kuma.io/install/latest/?utm_source=Kuma&utm_medium=Kuma-GUI"
                        type="button"
                      >
                        
                        <!---->
                        
                        
                         Update 
                        
                        <!---->
                      </a>
                    </div>
                  </div>
                  
                </div>
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex justify-between items-center"
      >
        <!--v-if-->
      </div>
    </header>
    <div
      class="main-content-container"
    >
      <aside
        class="has-subnav subnav-expanded"
        id="the-sidebar"
      >
        <div
          class="main-nav"
        >
          <div
            class="top-nav"
          >
            
            <div
              class="is-active is-menu-item nav-item"
              data-testid="home"
              iconcustom=""
              subnav="[object Object]"
            >
              <a>
                <div
                  class="nav-icon"
                >
                  
                  <!--v-if-->
                  
                </div>
                <!-- nav title separator -->
                
                <!-- nav link -->
                <div
                  class="nav-link"
                >
                  
                  Service Mesh 
                  <!--v-if-->
                  
                </div>
                
              </a>
            </div>
            
          </div>
          <div
            class="bottom-nav"
          >
            
            <div
              class="is-menu-item nav-item"
              data-testid="diagnostics"
            >
              <a>
                <div
                  class="nav-icon"
                >
                  
                  <span
                    class="kong-icon-gearFilled kong-icon"
                    data-v-637de6fa=""
                    height="18"
                    width="18"
                  >
                    <svg
                      height="18"
                      role="img"
                      viewBox="0 0 16 16"
                      width="18"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      
  
                      <title>
                        Settings
                      </title>
                      
  
                      <path
                        d="M8.63962039 0c.2152152 0 .40628463.13771505.47434165.34188612l.63997656 1.91853291c.3708976.11319954.7265407.26136452 1.0630881.4406537l1.8103874-.9038671c.1924944-.09624716.4249801-.05851992.5771602.09366021l.9045599.90455983c.1521801.15218013.1899073.38466585.0936602.57716018l-.904073 1.81000104c.1795192.33691975.3278441.69298066.44112 1.06432821l1.9182725.63912286C15.862285 6.95409498 16 7.14516441 16 7.36037961v1.27924078c0 .2152152-.137715.40628463-.3418861.47434165l-1.918532.63997355c-.1132072.37092351-.2613848.72659041-.4406907 1.06315871l.9039032 1.8103198c.0962471.1924944.0585199.4249801-.0936602.5771602l-.9045599.9045599c-.1521801.1521801-.3846658.1899073-.5771602.0936602l-1.8103171-.9039046c-.3365681.179306-.692235.3274837-1.0631584.4406912l-.63997656 1.9185329C9.04590502 15.862285 8.85483559 16 8.63962039 16H7.36037961c-.2152152 0-.40628463-.137715-.47434165-.3418861L6.2460614 13.739581c-.3708976-.1131996-.72654071-.2613646-1.06308807-.4406537l-1.81038748.9038671c-.19249433.0962471-.42498005.0585199-.57716018-.0936602l-.90455983-.9045599c-.15218013-.1521801-.18990737-.3846658-.09366021-.5771602l.90407299-1.810001c-.17951921-.3369197-.32784407-.6929806-.44112007-1.0643282L.34188612 9.11396204C.13771505 9.04590502 0 8.85483559 0 8.63962039V7.36037961c0-.2152152.13771505-.40628463.34188612-.47434165l1.918532-.63997355c.11320723-.37092352.26138479-.72659044.44069063-1.06315869l-.90390312-1.81031987c-.09624716-.19249433-.05851992-.42498005.09366021-.57716018l.90455983-.90455983c.15218013-.15218013.38466585-.18990737.57716018-.09366021l1.81031714.90390458c.33656815-.17930601.69223497-.32748375 1.06315841-.44069118L6.88603796.34188612C6.95409498.13771505 7.14516441 0 7.36037961 0h1.27924078zM8 6c-1.1045695 0-2 .8954305-2 2s.8954305 2 2 2 2-.8954305 2-2-.8954305-2-2-2z"
                        fill="var(--SidebarIconColor)"
                        fill-rule="evenodd"
                      />
                      

                    </svg>
                    

                  </span>
                  
                </div>
                <!-- nav title separator -->
                
                <!-- nav link -->
                <div
                  class="nav-link"
                >
                  
                  Diagnostics 
                  <!--v-if-->
                  
                </div>
                
              </a>
            </div>
            
          </div>
        </div>
        <div
          class="secondary-nav"
        >
          <div
            class="mt-3"
          >
            
            <div
              class="px-4 pb-4"
            >
              <div>
                <h3
                  class="menu-title uppercase"
                >
                   Filter by Mesh: 
                </h3>
                <select
                  class="mesh-selector"
                  id="mesh-selector"
                  name="mesh-selector"
                >
                  <option
                    value="all"
                  >
                     All Meshes 
                  </option>
                  
                  
                </select>
              </div>
            </div>
            
          </div>
          <!-- &lt;div
      class="arrow"
      @click="handleToggle"
    /&gt; -->
          <div
            class="subnav-title"
          >
            <span
              class="text-uppercase"
            >
              
              <a>
                Service Mesh
              </a>
              
            </span>
          </div>
          
          
          
          <div
            class="is-menu-item nav-item"
            data-testid="global-overview"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                Overview 
                <!--v-if-->
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item nav-item"
            data-testid="mesh-child"
            pathflip="true"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                Meshes 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item is-title nav-item"
            data-testid=""
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              <div
                class="title-text"
              >
                <span
                  class="text-uppercase"
                >
                  Zones
                </span>
              </div>
            </a>
          </div>
          <div
            class="is-menu-item nav-item"
            data-testid="zones"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                Zone CPs 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item nav-item"
            data-testid="zoneingresses"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                Zone Ingresses 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item nav-item"
            data-testid="zoneegresses"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                Zone Egresses 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item is-title nav-item"
            data-testid=""
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              <div
                class="title-text"
              >
                <span
                  class="text-uppercase"
                >
                  Services
                </span>
              </div>
            </a>
          </div>
          <div
            class="is-menu-item nav-item"
            data-testid="internal-services"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                Internal 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item nav-item"
            data-testid="external-services"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                External 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item is-title nav-item"
            data-testid=""
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              <div
                class="title-text"
              >
                <span
                  class="text-uppercase"
                >
                  Data plane proxies
                </span>
              </div>
            </a>
          </div>
          <div
            class="is-menu-item nav-item"
            data-testid="dataplanes"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                All 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item is-nested nav-item"
            data-testid="standard-dataplanes"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                Standard 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item is-nested nav-item"
            data-testid="gateway-dataplanes"
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              
              <!-- nav link -->
              <div
                class="nav-link"
              >
                
                Gateway 
                <span
                  class="amount amount--empty"
                >
                  0
                </span>
                
              </div>
              
            </a>
          </div>
          <div
            class="is-menu-item is-title nav-item"
            data-testid=""
          >
            <a>
              <!--v-if-->
              <!-- nav title separator -->
              <div
                class="title-text"
              >
                <span
                  class="text-uppercase"
                >
                  Policies
                </span>
              </div>
            </a>
          </div>
          
        </div>
      </aside>
      <main
        class="main-content"
      >
        <div>
          <!--v-if-->
          <!---->
        </div>
        <!--v-if-->
        <ul
          class="k-breadcrumbs"
          data-v-a87af1c0=""
        >
          
          <li
            class="k-breadcrumbs-item truncate"
            data-v-a87af1c0=""
          >
            <a
              class="no-underline"
              data-v-a87af1c0=""
            >
              <!---->
              <!---->
            </a>
            <span
              class="kong-icon-chevronRight kong-icon"
              data-v-637de6fa=""
              data-v-a87af1c0=""
            >
              <svg
                fill="var(--grey-500)"
                height="15"
                role="img"
                viewBox="0 0 20 20"
                width="15"
                xmlns="http://www.w3.org/2000/svg"
              >
                
  
                
  
                <path
                  d="m7.3 14.7 5-5.2-5-5.2"
                  fill="none"
                  stroke="var(--grey-500)"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                

              </svg>
              

            </span>
          </li>
          
        </ul>
        <transition-stub>
          <div>
             TestComponent 
          </div>
        </transition-stub>
      </main>
    </div>
  </div>
</div>
`;
